### PR TITLE
feat(tools): enhance grep search for streamingly reading files to reduce memory usage

### PR DESCRIPTION
**Pull request by bowenliang123**
## Description

[Describe what this PR does and why]
- Stream files line-by-line using open() iterator (vs read_text())
- Add sliding window (deque) for context management
- Add _emit_match_entries() and _output_context_for_hit() helpers
- Rename outputted_hits → emitted_hit_lines for clarity
- Fix off-by-one in output size checking (use > instead of >=)
- Check size before appending separator (not after)
- Add comprehensive test suite (33 tests)

**Related Issue:** Fixes #(issue_number) or Relates to #(issue_number)

**Security Considerations:** [If applicable, e.g. channel auth, env/config handling]

## Type of Change

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation
- [x] Refactoring

## Component(s) Affected

- [x] Core / Backend (app, agents, config, providers, utils, local_models)
- [ ] Console (frontend web UI)
- [ ] Channels (DingTalk, Feishu, QQ, Discord, iMessage, etc.)
- [ ] Skills
- [ ] CLI
- [ ] Documentation (website)
- [ ] Tests
- [ ] CI/CD
- [ ] Scripts / Deploy

## Checklist

- [x] I ran `pre-commit run --all-files` locally and it passes
- [x] If pre-commit auto-fixed files, I committed those changes and reran checks
- [x] I ran tests locally (`pytest` or as relevant) and they pass
- [x] Documentation updated (if needed)
- [x] Ready for review

## Testing

[How to test these changes]

## Local Verification Evidence

```bash
pre-commit run --all-files
# paste summary result

pytest
# paste summary result
```

## Additional Notes

[Optional: any other context]

### CODE REVIEW

**Comment by xieyxclack** [comment]
Thank you for your contribution! Please refer to the suggestions from Copilot and Gemini first, and make corrections if you find them helpful :)

**Comment by bowenliang123** [comment]
I'm new to the project. Let's see how I could fix the linting errors.

**Comment by xieyxclack** [comment]
@bowenliang123 Run `pre-commit run --all-files`

**Comment by bowenliang123** [comment]
Thanks for the hint.

Somehow, I've struggled with errors from errors of unrelated changes in local when running `pre-commit run --all-files`.
`error: cannot format src/copaw/agents/tools/file_search.py: module 'ast' has no attribute 'Str'`

image

**Comment by xieyxclack** [comment]
Is your base branch the "main" branch? It seems to work fine in my local tests. Or perhaps the way you installed pre-commit is incorrect?

Refer to https://github.com/agentscope-ai/CoPaw/blob/main/CONTRIBUTING.md#4-code-and-quality

**Comment by bowenliang123** [comment]
0. yes, main branch is my base branch
1. latest pre-commit v4.5.1 is installed and used
2. seems only the black tool works incorrectly
3. the error could be raised by black from Python , maybe related to Python version issue. However, .venv is inited and sync by uv tool and Python 3.12 is ready for use. 
Don't know how and what happen to the black tool itself.

**Comment by xieyxclack** [comment]
I can assist you with the formatting issues later; for now, let's focus on the actual changes :)

**Comment by bowenliang123** [comment]
Agree. The remaining issue is not printing the coming lines. I will try to fix it.

**Comment by xieyxclack** [comment]
I have a few suggestions to make this implementation complete:

- The current implementation provides context lines **before** the match (like `grep -B`), which is a clean and efficient design for streaming. However, the docstring at line 328 still describes "before and after" context (like `grep -C`). Please update the docstring to accurately reflect the current behavior.
- The deque buffer size should be `maxlen=context_lines + 1` (instead of `context_lines or 1`) to properly include both the context lines and the matched line

And I notice that gemini has gave simialr suggestion :)

**Comment by bowenliang123** [comment]
Hi @xieyxclack , kind thanks to the detailed review.

I've fixed the errors thrown when running`pre-commit run --all-files`. It turns out that the system-wide `pre-commit` is installed via brew, which makes it using Python 3.14. And the current version `black` is not compatible with Python 3.14. I followed the instruction again and reinit the venv env with `uv venv --python 3.12 --seed` to set python and pip version correctly.

I'm trying more appropriate approaches for the "before and after" context. The overlapping cases are challenging the read-only-once way.

**Comment by bowenliang123** [comment]
Hi @xieyxclack , it comes back with a refactored version using a sliding window and a read-once policy implemented.
And also, with the following improvements:
- pinned out the expected grep result into tests for `walk_and_grep`
- more frequent checking the cancelled state of the notification event for earlier getting out of the scanning

File: tests/unit/agents/tools/test_file_search.py [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+def test_walk_and_grep_single_file_mode_with_context(temp_dir):
+    """Test single-file mode with context_lines=2 and five hits demonstrating
+    four distinct gap sizes between consecutive hits.
+
+    File (14 lines):
+      1:  '1'
+      2:  '2 hit'   -- HIT 1
+      3:  '3 hit'   -- HIT 2 (gap=0, adjacent to HIT 1)
+      4:  '4'           (gap=1 between HIT 2 and 3: only line 4)
+      5:  '5 hit'   -- HIT 3
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply tests/unit/agents/tools/test_file_search.py:393]: resolved.

File: src/copaw/agents/tools/file_search.py [hunk 2]
```diff
@@ -216,47 +213,178 @@ def _walk_and_grep(  # noqa: C901  pylint: disable=too-many-branches,too-many-lo
...
+        )
 
-            start = max(0, line_no - 1 - context_lines)
-            end = min(len(lines), line_no + context_lines)
+        # Sliding window holds (line_no, line_content) tuples
+        window: list[tuple[int, str]] = []
+        # Indices in window where matches occurred, ordered by position
+        hit_indices: list[int] = []
+        # Track which line numbers have been output to avoid duplicates
+        outputted_hits: set[int] = set()
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: I don't recommend removing it. The outputted_hits set is not redundant.
The key is in the EOF processing logic (lines 371-385):
for hit_idx in hit_indices:           # Iterates, doesn't pop
    hit_line_no = window[hit_idx][0]
    if hit_line_no in outputted_hits:  # Deduplication check
        continue
Here, a for loop iterates over hit_indices without pop(0). If the same line number appears multiple times in hit_indices (e.g., the same line matches the regex multiple times, or gets accumulated across different sliding window positions), without outputted_hits it would be output multiple times.
The AI's suggestion is incomplete. It only considered the deduplication effect of pop(0) but missed that EOF processing uses iteration rather than pop to process remaining hits.
Keep outputted_hits.

File: src/copaw/agents/tools/file_search.py [hunk 3]
```diff
@@ -216,47 +213,178 @@ def _walk_and_grep(  # noqa: C901  pylint: disable=too-many-branches,too-many-lo
...
+                    if start_line <= ln <= end_line:
+                        # Check limits before appending each line to enforce caps strictly
+                        if len(matches) >= _MAX_MATCHES:
+                            return False
+                        prefix = ">" if idx == hit_idx else " "
+                        entry = f"{disp_path}:{ln}:{prefix} {content}"
+                        total_chars += len(entry) + 1
+                        if total_chars >= _MAX_OUTPUT_CHARS:
+                            return False
+                        matches.append(entry)
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: I don't recommend this change. The hit_idx is None branch is reachable.
The key point: _output_context_for_hit internally initializes hit_idx = None and then searches:
def _output_context_for_hit(hit_line_no, line_buffer, disp_path):
    hit_idx = None
    for idx, (ln, _) in enumerate(line_buffer):
        if ln == hit_line_no:
            hit_idx = idx
            break
    # hit_idx is used here
All callers pass only hit_line (the line number), not hit_idx. The function itself searches for the matching line number.
When the sliding window slides a matched line out of the buffer (e.g., for very large files), hit_idx will be None and the fallback scan is needed.
The AI's suggestion has two problems:
1. It references hit_idx, but the function signature doesn't have this parameter at all
2. The None branch is actually reachable, not dead code

File: src/copaw/agents/tools/file_search.py [hunk 4]
```diff
@@ -216,47 +213,178 @@ def _walk_and_grep(  # noqa: C901  pylint: disable=too-many-branches,too-many-lo
...
+        display_path = (
+            file_path.name
+            if single_file
+            else _relative_display(file_path, search_root)
+        )
 
-            start = max(0, line_no - 1 - context_lines)
-            end = min(len(lines), line_no + context_lines)
+        # Sliding window holds (line_no, line_content) tuples
+        window: list[tuple[int, str]] = []
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: OK, changed to use deque.

File: tests/unit/agents/tools/test_file_search.py [hunk 5]
```diff
@@ -0,0 +1,704 @@
+# -*- coding: utf-8 -*-
+"""Tests for file_search module — _walk_and_grep and _walk_and_glob."""
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply tests/unit/agents/tools/test_file_search.py]: Fixed.

File: src/copaw/agents/tools/file_search.py [hunk 6]
```diff
@@ -216,47 +214,178 @@ def _walk_and_grep(  # noqa: C901  pylint: disable=too-many-branches,too-many-lo
...
+                    if start_line <= ln <= end_line:
+                        # Check limits before appending each line to enforce caps strictly
+                        if len(matches) >= _MAX_MATCHES:
+                            return False
+                        prefix = ">" if ln == hit_line_no else " "
+                        entry = f"{disp_path}:{ln}:{prefix} {content}"
+                        total_chars += len(entry) + 1
+                        if total_chars >= _MAX_OUTPUT_CHARS:
+                            return False
+                        matches.append(entry)
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: OK. Updated handling for "truncated:".

File: src/copaw/agents/tools/file_search.py [hunk 7]
```diff
@@ -216,47 +214,205 @@ def _walk_and_grep(  # noqa: C901  pylint: disable=too-many-branches,too-many-lo
...
+            For each hit, we output lines from (hit - context_lines) to
+            (hit + context_lines), clamped to file boundaries. The hit line
+            is marked with '>' prefix, others with ' '.
+
+            Uses fast path when hit is found in buffer: direct slice indexing.
+            Falls back to range scan for edge cases (e.g., file start/end).
+            """
+            nonlocal total_chars
+            start_line = max(1, hit_line_no - context_lines)
+            end_line = hit_line_no + context_lines
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: `nonlocal total_chars` is still required, as line 327 has `total_chars += 4`.

File: src/copaw/agents/tools/file_search.py [hunk 8]
```diff
@@ -216,47 +214,205 @@ def _walk_and_grep(  # noqa: C901  pylint: disable=too-many-branches,too-many-lo
...
+            nonlocal total_chars
+
+            for ln, content, is_hit in entries:
+                if len(matches) >= _MAX_MATCHES:
+                    return False
+                prefix = ">" if is_hit else " "
+                entry = f"{disp_path}:{ln}:{prefix} {content}"
                 total_chars += len(entry) + 1
+                if total_chars >= _MAX_OUTPUT_CHARS:
+                    return False
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: Resolved.

File: src/copaw/agents/tools/file_search.py [hunk 9]
```diff
@@ -216,47 +331,123 @@ def _walk_and_grep(  # noqa: C901  pylint: disable=too-many-branches,too-many-lo
...
 
-        for line_no, line in enumerate(lines, start=1):
-            if not regex.search(line):
-                continue
+        # Sliding window holds (line_no, line_content) tuples
+        window: deque[tuple[int, str]] = deque(maxlen=window_size)
+        # Indices in window where matches occurred, ordered by position
+        hit_indices: list[int] = []
+        # Track which line numbers have been output to avoid duplicates
+        outputted_hits: set[int] = set()
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: Thanks. Resolved.

File: src/copaw/agents/tools/file_search.py [hunk 10]
```diff
@@ -162,6 +161,123 @@ def _resolve_search_root(
...
+        were appended, False if limits were reached.
+    """
+    for ln, content, is_hit in entries:
+        if len(matches) >= _MAX_MATCHES:
+            return False, total_chars
+        prefix = ">" if is_hit else " "
+        entry = f"{disp_path}:{ln}:{prefix} {content}"
+        total_chars += len(entry) + 1
+        if total_chars >= _MAX_OUTPUT_CHARS:
+            return False, total_chars
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: Thanks. Fixed.

File: src/copaw/agents/tools/file_search.py [hunk 11]
```diff
@@ -162,6 +161,124 @@ def _resolve_search_root(
...
+        return False, total_chars
+
+    # Append separator if needed
+    if context_lines > 0:
+        if len(matches) >= _MAX_MATCHES:
+            return False, total_chars
+        matches.append("---")
+        total_chars += 4
+        if total_chars >= _MAX_OUTPUT_CHARS:
+            return False, total_chars
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply src/copaw/agents/tools/file_search.py]: OK, Fixed.

**Comment by bowenliang123** [comment]
Hi @xieyxclack , it should be good to go, after resolving several rounds' review from Github Copilot and Genmini. Please  check.

**Comment by bowenliang123** [comment]
> * **Remove `emitted_hit_lines` set** - it seems unnecessary and redundant?
> * **Remove unreachable fallback path** - dead code that will never execute
> * **Remove redundant `if start_line <= ln <= end_line` check** - Already constrained by slice

@xieyxclack  You're right. These three defensive approaches are not in need. I've removed them.

As for the test files, they are quite important to pin down the expected cases, especially in corner cases like overlapping, repeated hits etc., considering it's a challenge in the read-once strategy. Are you sure to remove it?

**Comment by bowenliang123** [comment]
> ## Welcome to CoPaw! 🎉
> Thank you @bowenliang123 for your first contribution! Your PR has been merged. 🚀
> 
> We'd love to give you a shout-out in our release notes! If you're comfortable sharing, please reply to this comment with your social media handles using the format below:
> 
> ```
> discord: your_discord_handle
> x: your_x_handle
> xiaohongshu: your_xiaohongshu_id
> ```
> 
> > **Note:** Only share what you're comfortable with — all fields are optional.
> 
> Thanks again for helping make CoPaw better!

discord: bowenliang_74459_95085
x: bowenliang123
xiaohongshu: 5072567189
